### PR TITLE
[BIOMAGE-1184] - Prevent unnecessary re-runs of GEM2S

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1184

Inconsistent re-run behaviour was caused because of inconsistent metadata behaviour. 
- Creating a new metadata in the UI inserts the new metadata key into the `project` slice, and inserts default values for this key for each sample (in `samples` slice). This key is saved to DynamoDB for `project`, but the default values that are inserted for each sample are not saved to DynamoDB.
- When GEM2S is run, if a metadata key exists, but the sample does not have any entry for that key, the API fills in the missing values a default value. However, this value does not actually exist in DynamoDB. Therefore, when Data Management is reloaded, no metadata value is loaded for samples. By default, the UI shows `N.A.` for missing metadata value, even if the value is not there.
-  Because the metadata does not exist in the samples slice (even if it is showing `N.A` in the UI), the `Launch Analysis` button will be disabled. This is really confusing for users because they see metadata is there, but the `Launch Analysis` button isn't clickable. The user then have to delete the metadata key with empty values to enable the `Launch Analysis` button. Removing the metadata key results in a different project configuration, which causes a re-run of GEM2S.

#### Link to staging deployment URL 

https://ui-agi-ui489-api222.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

API - https://github.com/biomage-ltd/api/pull/222
Integration - https://github.com/biomage-ltd/testing/pull/21

#### Anything else the reviewers should know about the changes here

To address the issues above : 
- Save metadata to database after adding new metadata key
- Provide a tooltip on the Launch Analysis button to show the user what's keeping the button from becoming available.

This PR also :
- Fix metadata creation problem. When creating metadata, the new metadata is added to all samples, regardless of project. This PR introduces a fix so that metadata is added only to the relevant project.
- Changes the name of the Launch Analysis button to be indicative of whether the user is going to run Gem2S (Process project) or not (Go to Data Processing).
- Remove the launch analysis modal

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging
- [x] Ran integration test

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [x] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/redux/actions/projects/createMetadataTrack.js
Comment (aerlaut) [inline src/redux/actions/projects/createMetadataTrack.js:38]: This will take all samples (Including samples which are not in the project)

File: src/redux/actions/projects/createMetadataTrack.js
Comment (aerlaut) [inline src/redux/actions/projects/createMetadataTrack.js:39]: This will only add metadata to samples that are in the project

File: src/components/data-management/ProjectMenu.jsx
Comment (cosa65) [inline src/components/data-management/ProjectMenu.jsx]: Can we create and use `gem2sNotSuccessful` and `ProjectHashNotEqual` in this function directly?

File: src/components/data-management/ProjectMenu.jsx
Comment (cosa65) [inline src/components/data-management/ProjectMenu.jsx]: `ProjectHashNotEqual` should be `projectHashNotEqual`, right?

File: src/components/data-management/ProjectMenu.jsx
Comment (cosa65) [inline src/components/data-management/ProjectMenu.jsx]: Why is the hash initialized for each project here? Why not calculate it in the reducer when project is loaded/modified or something? And then we don't need to recalculate whenever this is rerendered

Comment (aerlaut) [reply src/components/data-management/ProjectMenu.jsx:73]: Hmm.. Good idea. I'll move it to the reducers,

**Comment by ivababukova** [comment]
I am afraid this will need to wait until I merge in https://github.com/biomage-ltd/ui/pull/494 and then it will need to be rewritten a bit. I have rewritten that part of the UI components, now ProjectDetails is completely different

**Review by marcellp — DISMISSED** [review]
lgtm

File: src/redux/reducers/samples/initialState.js
Comment (aerlaut) [inline src/redux/reducers/samples/initialState.js:19]: This was added to convert `fileNames` which is a Set into an array for JSON. Filenames is also stored as a list in DynamoDB.

File: src/redux/reducers/samples/samplesFileUpdate.js
Comment (aerlaut) [inline src/redux/reducers/samples/samplesFileUpdate.js:23]: The main reason we were using a set was to ensure that there is no duplicate name. If we're storing `fileNames` as a set, we don't have to convert it back to an array.

File: src/redux/reducers/samples/samplesFileUpdate.js
Comment (aerlaut) [inline src/redux/reducers/samples/samplesFileUpdate.js]: We can do name checking with array also. Compared to the previous implementation, it seems to be cheaper because we only create new array if the current file name doesn't exist.

**Comment by marcellp** [comment]
What's the status on this @aerlaut? Can we merge it in at this state?

**Comment by aerlaut** [comment]
This PR is ready to merge

File: src/components/data-management/ProjectMenu.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectMenu.jsx]: small thing: you can call this `DownloadDataButton` so that it is exactly like `LaunchAnalysisButton`.

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (ivababukova) [inline src/components/data-management/LaunchAnalysisButton.jsx]: small thing: you can rename `AnalysisModal` to `LaunchAnalysisModal` so that its name follows the same principle like `FileUploadModal`, `LaunchAnalysisButton`, `DownloadButton`

File: src/redux/actions/pipeline/runGem2s.js
Comment (ivababukova) [inline src/redux/actions/pipeline/runGem2s.js]: why passing this at all? Can we simply assume that we only call the API if we have to run gem2s, otherwise we don't? What is the point of calling the API if we don't need to run gem2s? Why does the API need to know about this gem2s hash?

File: src/redux/actions/pipeline/runGem2s.js
Comment (ivababukova) [inline src/redux/actions/pipeline/runGem2s.js]: In what cases will we have not defined hash value?

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (ivababukova) [inline src/components/data-management/LaunchAnalysisButton.jsx]: what is the point of `existingParamsHash && existingParamsHash`?

Comment (aerlaut) [reply src/components/data-management/LaunchAnalysisButton.jsx]: `existingParamsHash` doesn't exist if the project has never been run. So doing `existingParamsHash && ...` will evaluate to false as `existingParamasHash` will be `undefined`. It can be made clearer by setting a default value on line 55. I thnink I'll do that.

File: src/components/data-management/ProjectMenu.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectMenu.jsx]: changed

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (aerlaut) [reply src/components/data-management/LaunchAnalysisButton.jsx]: changed

File: src/redux/actions/pipeline/runGem2s.js
Comment (aerlaut) [reply src/redux/actions/pipeline/runGem2s.js]: When we `reRunGems()` from the GEM2S loading screen. The user can not edit the project again, but we need GEM2S to re-run with the same params. 

That said, this conversation uncovers a bug. In this case, we'd like GEM2S to run again, but we still want to still use the old GEM2S params. So I thinkn it's better to let `gem2sHash` be null for the case that we want GEM2S to run again, but use the `oldGem2sHash` to save.

File: src/redux/actions/pipeline/runGem2s.js
Comment (aerlaut) [reply src/redux/actions/pipeline/runGem2s.js]: You're right. In the cases where we want to run GEM2S, we still need to pass the hash. I'll change it accordingly.

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (aerlaut) [inline src/components/data-management/LaunchAnalysisButton.jsx]: remove this

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (aerlaut) [inline src/components/data-management/LaunchAnalysisButton.jsx]: Refactor this functional body

File: src/components/data-management/LaunchAnalysisModal.jsx
Comment (aerlaut) [inline src/components/data-management/LaunchAnalysisModal.jsx]: Remove this component

File: src/components/data-management/ProjectMenu.jsx
Comment (aerlaut) [inline src/components/data-management/ProjectMenu.jsx]: MIstake, revert

File: src/components/data-management/ProjectMenu.jsx
Comment (aerlaut) [inline src/components/data-management/ProjectMenu.jsx:13]: Has to be added to add the itnegration test constants

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (aerlaut) [reply src/components/data-management/LaunchAnalysisButton.jsx]: refactored

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (aerlaut) [reply src/components/data-management/LaunchAnalysisButton.jsx]: removed

File: src/__test__/redux/actions/samples/saveSamples.test.js
Comment (ivababukova) [inline src/__test__/redux/actions/samples/saveSamples.test.js]: rename to "correctly update existing samples in database"

**Review by ivababukova — DISMISSED** [review]

**Comment by StefanBabukov** [comment]
When i open data-management I get this - 
image
Have you seen this?

**Comment by StefanBabukov** [comment]
Might be that the experiment i copied has corrupted values, not from your changes

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (StefanBabukov) [inline src/components/data-management/LaunchAnalysisButton.jsx]: Can you not use one single experimentId variable across this file instead of defining a new one here and on line 96?  Also that way `calculateGem2sRerunStatus` won't need an experimentId argument.

File: src/redux/actions/projects/createMetadataTrack.js
Comment (ivababukova) [inline src/redux/actions/projects/createMetadataTrack.js:65]: @cosa65 you and Agi should chat about this line

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (aerlaut) [reply src/components/data-management/LaunchAnalysisButton.jsx]: yup, that can be done 👍

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/LaunchAnalysisButton.test.jsx]: Rename `Launch analysis` to `Process project`.

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/LaunchAnalysisButton.test.jsx]: Rename `Launch analysis` to `Process project`.

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/LaunchAnalysisButton.test.jsx]: Rename `Launch analysis` to `Process project`.

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/LaunchAnalysisButton.test.jsx:160]: These look like some thorough tests Agi, very nicely done!

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/LaunchAnalysisButton.test.jsx]: Renamed

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/LaunchAnalysisButton.test.jsx]: Renamed

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/LaunchAnalysisButton.test.jsx]: Renamed

File: src/__test__/components/data-management/LaunchAnalysisButton.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/LaunchAnalysisButton.test.jsx:160]: Thanks Iva!

**Comment by aerlaut** [comment]
Merging this PR will cause integration tests to fail. The integration tests look for a button with the words "Launch Analysis" in it. This PR changes that into "Process project" or "Go to Data Processing". I've opened a ticket to fix this : https://biomage.atlassian.net/browse/BIOMAGE-1463

File: package.json
Comment (aerlaut) [inline package.json]: I had to add this as tests couldn't run. It says it needs the required package. Did that happen with you too @cosa65?

**Review by ivababukova — DISMISSED** [review]

**Review by StefanBabukov — APPROVED** [review]

**Pull request merged**
